[1] [contrib/engines] setup typing in contrib part of the library

RaviTezu opened the pull request:
Fixes #977

Description: Setup Typing in `contrib` part of the library

RaviTezu commented:
Hi @vfdev-5 Could you please take a look when you have a minute ? 
I see, the `(ubuntu-latest, 3.5, pytorch)` checks are failing as type annotations are not supported in Py3.5?

vfdev-5 commented on ignite/contrib/engines/common.py:
```diff
@@ -20,29 +22,32 @@
+    stop_on_nan: Optional[bool] = True,
+    clear_cuda_cache: Optional[bool] = True,
+    save_handler: Optional[Union[Callable, BaseSaveHandler]] = None,
+    **kwargs: Mapping,
```
Let's set kwargs and args as Any.
```suggestion
    **kwargs: Any
```
Please, use this convention for all other functions.

RaviTezu replied:
Done

vfdev-5 commented on ignite/contrib/engines/common.py:
```diff
@@ -200,21 +205,21 @@ def output_transform(x, index, name):
+    stop_on_nan: Optional[bool] = True,
+    clear_cuda_cache: Optional[bool] = True,
+    save_handler: Optional[Union[Callable, BaseSaveHandler]] = None,
+    **kwargs: Mapping,
```
```suggestion
    **kwargs: Any
```

vfdev-5 commented on ignite/contrib/engines/common.py:
```diff
@@ -20,29 +22,32 @@
+    save_every_iters: Optional[int] = 1000,
+    output_path: Optional[str] = None,
+    lr_scheduler: Optional[Union[ParamScheduler, _LRScheduler]] = None,
+    with_gpu_stats: Optional[bool] = False,
```
I think `Optional` is only for kwargs that are None. So here it is : 
 ```suggestion
    with_gpu_stats: bool = False,
```

Could you please fix for all other kwargs.

RaviTezu replied:
Done

vfdev-5 commented on ignite/contrib/engines/common.py:
```diff
@@ -257,10 +262,14 @@ def setup_any_logging(logger, logger_module, trainer, optimizers, evaluators, lo
+def _setup_logging(
+    logger: BaseLogger,
+    trainer: Engine,
+    optimizers: Optional[Union[Optimizer, Dict[str, Optimizer]]],
```
this line and 2 below ones are not Optional

RaviTezu replied:
Done

vfdev-5 commented on ignite/contrib/engines/tbptt.py:
```diff
@@ -48,7 +59,7 @@ def create_supervised_tbptt_trainer(
         loss_fn (torch.nn loss function): the loss function to use.
         tbtt_step (int): the length of time chunks (last one may be smaller).
-        dim (int): axis representing the time dimension.
+        dim (int, optional): axis representing the time dimension.
```
Let's set it as before (not optional)

RaviTezu replied:
Done

vfdev-5 commented on ignite/contrib/engines/tbptt.py:
```diff
@@ -73,7 +84,7 @@ def create_supervised_tbptt_trainer(
     """
 
-    def _update(engine, batch):
+    def _update(engine: Engine, batch: Tuple[torch.Tensor, torch.Tensor]):
```
Let's use `Sequence[torch.Tensor]` for batch

RaviTezu replied:
Done

vfdev-5 commented on ignite/contrib/handlers/mlflow_logger.py:
```diff
@@ -1,14 +1,125 @@
+
+        mlflow.end_run()
+
+    def _create_output_handler(self, *args: Any, **kwargs: Mapping):
```
```suggestion
    def _create_output_handler(self, *args: Any, **kwargs: Any):
```

RaviTezu replied:
Done

vfdev-5 reviewed (commented):
Thanks a lot @RaviTezu for the PR !

I left some comments to improve it. Could you please apply those suggestions to all the files.

vfdev-5 commented on ignite/contrib/handlers/tqdm_logger.py:
```diff
@@ -273,13 +198,92 @@ def attach(
+        description: str,
+        metric_names: Optional[str] = None,
+        output_transform: Optional[Callable] = None,
+        closing_event_name: Any = Events.EPOCH_COMPLETED,
```
```suggestion
        closing_event_name: EventEnum = Events.EPOCH_COMPLETED,
```

RaviTezu replied:
Done

vfdev-5 commented on ignite/contrib/metrics/precision_recall_curve.py:
```diff
@@ -23,7 +25,7 @@ class PrecisionRecallCurve(EpochMetric):
             form expected by the metric. This can be useful if, for example, you have a multi-output model and
             you want to compute the metric with respect to one of the outputs.
-        check_compute_fn (bool): Optional default False. If True, `precision_recall_curve
+        check_compute_fn (bool, optional): Optional default False. If True, `precision_recall_curve
```
```suggestion
        check_compute_fn (bool): Default False. If True, `precision_recall_curve
```

vfdev-5 commented on ignite/contrib/metrics/roc_auc.py:
```diff
@@ -70,7 +72,7 @@ class RocCurve(EpochMetric):
             form expected by the metric. This can be useful if, for example, you have a multi-output model and
             you want to compute the metric with respect to one of the outputs.
-        check_compute_fn (bool): Optional default False. If True, `sklearn.metrics.roc_curve
+        check_compute_fn (bool, optional): Optional default False. If True, `sklearn.metrics.roc_curve
```
```suggestion
        check_compute_fn (bool): Default False. If True, `sklearn.metrics.roc_curve
```

vfdev-5 commented on ignite/contrib/metrics/roc_auc.py:
```diff
@@ -34,7 +36,7 @@ class ROC_AUC(EpochMetric):
             form expected by the metric. This can be useful if, for example, you have a multi-output model and
             you want to compute the metric with respect to one of the outputs.
-        check_compute_fn (bool): Optional default False. If True, `roc_curve
+        check_compute_fn (bool, optional): Optional default False. If True, `roc_curve
```
```suggestion
        check_compute_fn (bool): Default False. If True, `roc_curve
```

vfdev-5 reviewed (commented):
Thanks a lot @RaviTezu ! You did a huge work on typing whole contrib module.
Maybe, it would be better to split the PR into several ones...
I left several comments to improve it.

RaviTezu commented:
@vfdev-5 Sure, I will split this into multiple PRs.

RaviTezu commented on ignite/contrib/engines/common.py:
```diff
@@ -20,29 +22,32 @@
+    stop_on_nan: bool = True,
+    clear_cuda_cache: bool = True,
+    save_handler: Optional[Union[Callable, BaseSaveHandler]] = None,
+    **kwargs: Any,
```
I wish, `black` can do this for me :)

RaviTezu replied:
Done

RaviTezu commented on ignite/contrib/engines/common.py:
```diff
@@ -467,16 +519,23 @@ def setup_trains_logging(trainer, optimizers=None, evaluators=None, log_every_it
-    save_handler, evaluator, models, metric_name, n_saved=3, trainer=None, tag="val", **kwargs
+    save_handler: Union[Callable, BaseSaveHandler],
+    evaluator: Engine,
+    models: torch.nn.Module,
```
Done

RaviTezu commented on ignite/contrib/engines/tbptt.py:
```diff
@@ -1,6 +1,10 @@
 
 import torch
+from torch.nn import Module
+from torch.nn.modules.loss import _Loss
```
Done

RaviTezu commented on ignite/contrib/engines/tbptt.py:
```diff
@@ -1,6 +1,10 @@
+from typing import Callable, Mapping, Optional, Sequence, Tuple, Union
 
 import torch
+from torch.nn import Module
```
Done

RaviTezu commented on ignite/contrib/engines/tbptt.py:
```diff
@@ -27,7 +31,14 @@ def _detach_hidden(hidden):
-    model, optimizer, loss_fn, tbtt_step, dim=0, device=None, non_blocking=False, prepare_batch=_prepare_batch
+    model: Module,
+    optimizer: Optimizer,
+    loss_fn: _Loss,
```
Done

RaviTezu commented on ignite/contrib/handlers/param_scheduler.py:
```diff
@@ -137,7 +142,7 @@ def get_param(self) -> Union[List[float], float]:
 
     @classmethod
-    def simulate_values(cls, num_events, **scheduler_kwargs):
+    def simulate_values(cls, num_events: int, **scheduler_kwargs: Mapping):
```
ah, missed it. Done.

RaviTezu commented on ignite/contrib/handlers/param_scheduler.py:
```diff
@@ -513,7 +518,7 @@ def state_dict(self):
         return state_dict
 
-    def load_state_dict(self, state_dict):
+    def load_state_dict(self, state_dict: Dict[str, Any]):
```
Done. I made these changes in other places as well.

RaviTezu commented on ignite/contrib/handlers/param_scheduler.py:
```diff
@@ -1041,7 +1048,7 @@ def state_dict(self):
         return state_dict
 
-    def load_state_dict(self, state_dict):
+    def load_state_dict(self, state_dict: Dict):
```
Done.

RaviTezu commented on ignite/contrib/handlers/param_scheduler.py:
```diff
@@ -173,7 +178,7 @@ def simulate_values(cls, num_events, **scheduler_kwargs):
 
     @classmethod
-    def plot_values(cls, num_events, **scheduler_kwargs):
+    def plot_values(cls, num_events: int, **scheduler_kwargs: Mapping):
```
Done.

RaviTezu commented on ignite/contrib/metrics/average_precision.py:
```diff
@@ -22,7 +24,7 @@ class AveragePrecision(EpochMetric):
             form expected by the metric. This can be useful if, for example, you have a multi-output model and
             you want to compute the metric with respect to one of the outputs.
-        check_compute_fn (bool): Optional default False. If True, `average_precision_score
+        check_compute_fn (bool, optional): Optional default False. If True, `average_precision_score
```
Done

RaviTezu commented:
@vfdev-5 I have addressed your comments on the PR and pushed the changes. 
Let's split this PR based on the directories in `contrib` ? We can do `contrib/engines` in this PR and I will raise separate PRs for `contrib/handlers` and `contrib/metrics`. Works?

vfdev-5 commented:
Sounds good, thanks !

On Mon, Oct 5, 2020, 19:07 RaviTeja Pothana <notifications@github.com>
wrote:

> @vfdev-5 <https://github.com/vfdev-5> I have addressed your comments on
> the PR and pushed the changes.
> Let's split this PR based on the directories in contrib ? We can do
> contrib/engines in this PR and I will raise separate PRs for
> contrib/handlers and contrib/metrics. Works?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/ignite/pull/1351#issuecomment-703764875>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASYOH4NBSQFL2MNLNKNZK3SJH4O5ANCNFSM4SDQC3LQ>
> .
>

RaviTezu commented:
@vfdev-5 raised #1362  and #1363

vfdev-5 commented:
@RaviTezu python 3.5 still fails: https://github.com/pytorch/ignite/pull/1351/checks?check_run_id=1210448763#step:5:64
Could you please take a look.
Thanks for multiple PRs

RaviTezu commented:
@vfdev-5 Removed other files from this PR.

vfdev-5 commented on ignite/contrib/engines/common.py:
```diff
@@ -316,14 +332,20 @@ def setup_tb_logging(output_path, trainer, optimizers=None, evaluators=None, log
         **kwargs: optional keyword args to be passed to construct the logger.
 
     Returns:
-        :class:`~ignite.contrib.handlers.tensorboard_logger.TensorboardLogger`
```
Let's keep all those links as links, otherwise url is missing and the text can not be clicked on

vfdev-5 commented on ignite/contrib/engines/common.py:
```diff
@@ -340,14 +362,20 @@ def setup_visdom_logging(trainer, optimizers=None, evaluators=None, log_every_it
         **kwargs: optional keyword args to be passed to construct the logger.
 
     Returns:
-        :class:`~ignite.contrib.handlers.visdom_logger.VisdomLogger`
```
Same here

vfdev-5 commented on ignite/contrib/engines/common.py:
```diff
@@ -501,7 +560,7 @@ def gen_save_best_models_by_val_score(
         **kwargs: optional keyword args to be passed to construct :class:`~ignite.handlers.checkpoint.Checkpoint`.
 
     Returns:
-        A :class:`~ignite.handlers.Checkpoint` handler.
```
Same here, let's keep as it was

vfdev-5 commented on ignite/contrib/engines/common.py:
```diff
@@ -547,7 +611,7 @@ def save_best_model_by_val_score(
         **kwargs: optional keyword args to be passed to construct :class:`~ignite.handlers.checkpoint.Checkpoint`.
 
     Returns:
-        A :class:`~ignite.handlers.Checkpoint` handler.
```
Same here

vfdev-5 commented on ignite/contrib/engines/common.py:
```diff
@@ -573,7 +637,7 @@ def add_early_stopping_by_val_score(patience, evaluator, trainer, metric_name):
             `evaluator.state.metrics`.
 
     Returns:
-        A :class:`~ignite.handlers.EarlyStopping` handler.
```
Same here

vfdev-5 reviewed (commented):
Thanks for the update @RaviTezu !
I left few comment to improve the PR.

And looking docs preview: https://deploy-preview-1351--pytorch-ignite-preview.netlify.app/contrib/engines.html#module-ignite.contrib.engines.common
Docs now looks a bit urgly, do you know if we can do something to improve that ?
<img width="709" alt="Screen Shot 2020-10-06 at 10 01 14" src="https://user-images.githubusercontent.com/2459423/95175255-82d1d300-07bb-11eb-97be-0cddf31a7d0d.png">

RaviTezu commented:
@vfdev-5 I think, I didn't make any of the doc string changes. Do we have any CI job which does it automatically? I am not sure, if it was my IDE which did it. Reverted those changes and pushed just now.

vfdev-5 commented:
> Reverted those changes and pushed just now.

@RaviTezu thanks !

> Do we have any CI job which does it automatically?

We have only autopep8 fix github action but it should not touch the docstring...

RaviTezu commented:
@vfdev-5 Could you please check this PR now?

vfdev-5 reviewed (approved):
Thanks for the PR @RaviTezu !
LGTM !

RaviTezu commented:
Thanks a lot for the reviews @vfdev-5
